File: packages/protocol/src/actions.ts
import { Capability } from './capabilities';
import {
  DisplayInfo,
  FileItem,
  ForegroundAppState,
  MediaSessionState,
  VirtualRoot,
} from './messages';
import { PanelDefinition } from './panels';

export type Action =
  | {
      type: 'pointer.button';
      button: 'left' | 'right' | 'middle';
      state: 'down' | 'up' | 'click' | 'double_click';
    }
  | {
      type: 'pointer.scroll';
      dx: number;
      dy: number;
    }
  | {
      type: 'keyboard.key';
      key: string;
      state: 'down' | 'up' | 'tap';
      modifiers?: Array<'ctrl' | 'alt' | 'shift' | 'win'>;
    }
  | {
      type: 'keyboard.text';
      text: string;
    }
  | {
      type: 'keyboard.shortcut';
      keys: string[];
    }
  | {
      type: 'media.command';
      action:
        | 'play'
        | 'pause'
        | 'play_pause'
        | 'next'
        | 'prev'
        | 'volume_up'
        | 'volume_down'
        | 'set_volume'
        | 'mute'
        | 'seek';
      value?: number | null;
    }
  | {
      type: 'presentation.command';
      action: 'next' | 'prev' | 'start' | 'stop' | 'black_screen' | 'goto_slide';
      slideIndex?: number | null;
    }
  | {
      type: 'apps.launch';
      appId: string;
    }
  | {
      type: 'windows.action';
      windowId: string;
      action:
        | 'focus'
        | 'minimize'
        | 'maximize'
        | 'restore'
        | 'close'
        | 'snap_left'
        | 'snap_right'
        | 'move_to_display';
      targetDisplay?: number | null;
    }
  | {
      type: 'clipboard.set';
      text: string;
    }
  | {
      type: 'power.command';
      action: 'lock' | 'sleep' | 'restart' | 'shutdown';
    }
  | {
      type: 'macro.execute';
      macroId: string;
    };

export interface ActionResult {
  actionId: string;
  status: 'ok' | 'error';
  error?: string | null;
}

export type Query =
  | { type: 'files.list_roots' }
  | { type: 'files.browse'; rootId: string; subpath?: string }
  | { type: 'files.read_file'; rootId: string; subpath: string }
  | { type: 'panels.list' }
  | { type: 'clipboard.get' };

export type QueryResult<Q extends Query> = Q extends { type: 'files.list_roots' }
  ? VirtualRoot[]
  : Q extends { type: 'files.browse' }
    ? FileItem[]
    : Q extends { type: 'files.read_file' }
      ? { rootId: string; subpath: string; filename: string; contentBase64: string; size: number }
      : Q extends { type: 'panels.list' }
        ? PanelDefinition[]
        : Q extends { type: 'clipboard.get' }
          ? { text: string; timestamp: number }
          : unknown;

export type CompanionEvent =
  | { type: 'state.foreground_app'; data: ForegroundAppState }
  | { type: 'state.media_session'; data: MediaSessionState }
  | { type: 'state.displays'; data: DisplayInfo[] }
  | { type: 'state.clipboard'; data: { text: string; timestamp: number } }
  | { type: 'state.system_status'; data: { cpuUsagePercent?: number; ramUsagePercent?: number } };

export function getRequiredCapabilityForAction(action: Action): Capability {
  switch (action.type) {
    case 'pointer.button':
    case 'pointer.scroll':
      return 'input.mouse';
    case 'keyboard.key':
    case 'keyboard.text':
    case 'keyboard.shortcut':
      return 'input.keyboard';
    case 'media.command':
      return 'media.control';
    case 'presentation.command':
      return 'presentation.control';
    case 'apps.launch':
      return 'apps.launch';
    case 'windows.action':
      return 'windows.control';
    case 'clipboard.set':
      return 'clipboard.write';
    case 'power.command':
      switch (action.action) {
        case 'lock':
          return 'power.lock';
        case 'sleep':
          return 'power.sleep';
        case 'restart':
          return 'power.restart';
        case 'shutdown':
          return 'power.shutdown';
      }
      break;
    case 'macro.execute':
      return 'automation.execute';
  }
}

export function validateActionBounds(action: Action): { valid: boolean; error?: string } {
  switch (action.type) {
    case 'pointer.scroll':
      if (
        !Number.isFinite(action.dx) ||
        !Number.isFinite(action.dy) ||
        Number.isNaN(action.dx) ||
        Number.isNaN(action.dy)
      ) {
        return { valid: false, error: 'Scroll deltas must be finite numbers' };
      }
      if (Math.abs(action.dx) > 10000 || Math.abs(action.dy) > 10000) {
        return { valid: false, error: 'Scroll deltas exceed allowable limit' };
      }
      break;

    case 'keyboard.text':
      if (typeof action.text !== 'string' || action.text.length > 65536) {
        return { valid: false, error: 'Text payload exceeds allowable 64KB limit' };
      }
      break;

    case 'clipboard.set':
      if (typeof action.text !== 'string' || action.text.length > 1048576) {
        return { valid: false, error: 'Clipboard text exceeds allowable 1MB limit' };
      }
      break;

    case 'apps.launch':
      if (!action.appId || action.appId.trim().length === 0 || action.appId.length > 256) {
        return { valid: false, error: 'Invalid appId length' };
      }
      break;

    case 'windows.action':
      if (!action.windowId || action.windowId.trim().length === 0) {
        return { valid: false, error: 'Invalid windowId' };
      }
      if (
        action.targetDisplay !== undefined &&
        action.targetDisplay !== null &&
        (!Number.isInteger(action.targetDisplay) ||
          action.targetDisplay < 0 ||
          action.targetDisplay > 64)
      ) {
        return { valid: false, error: 'Invalid target display index' };
      }
      break;
  }

  return { valid: true };
}

File: crates/remote-protocol/src/actions.rs
use crate::capabilities::Capability;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum Action {
    #[serde(rename = "pointer.button")]
    PointerButton { button: String, state: String },
    #[serde(rename = "pointer.scroll")]
    PointerScroll { dx: f32, dy: f32 },
    #[serde(rename = "keyboard.key")]
    KeyboardKey {
        key: String,
        state: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        modifiers: Option<Vec<String>>,
    },
    #[serde(rename = "keyboard.text")]
    KeyboardText { text: String },
    #[serde(rename = "keyboard.shortcut")]
    KeyboardShortcut { keys: Vec<String> },
    #[serde(rename = "media.command")]
    MediaCommand {
        action: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        value: Option<f32>,
    },
    #[serde(rename = "presentation.command")]
    PresentationCommand {
        action: String,
        #[serde(
            rename = "slideIndex",
            default,
            skip_serializing_if = "Option::is_none"
        )]
        slide_index: Option<u32>,
    },
    #[serde(rename = "apps.launch")]
    AppsLaunch {
        #[serde(rename = "appId")]
        app_id: String,
    },
    #[serde(rename = "windows.action")]
    WindowsAction {
        #[serde(rename = "windowId")]
        window_id: String,
        action: String,
        #[serde(
            rename = "targetDisplay",
            default,
            skip_serializing_if = "Option::is_none"
        )]
        target_display: Option<u32>,
    },
    #[serde(rename = "clipboard.set")]
    ClipboardSet { text: String },
    #[serde(rename = "power.command")]
    PowerCommand { action: String },
    #[serde(rename = "macro.execute")]
    MacroExecute {
        #[serde(rename = "macroId")]
        macro_id: String,
    },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ActionResult {
    #[serde(rename = "actionId")]
    pub action_id: String,
    pub status: String, // "ok" | "error"
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

impl ActionResult {
    pub fn ok(action_id: impl Into<String>) -> Self {
        Self {
            action_id: action_id.into(),
            status: "ok".to_string(),
            error: None,
        }
    }

    pub fn error(action_id: impl Into<String>, msg: impl Into<String>) -> Self {
        Self {
            action_id: action_id.into(),
            status: "error".to_string(),
            error: Some(msg.into()),
        }
    }
}

impl Action {
    pub fn required_capability(&self) -> Capability {
        match self {
            Action::PointerButton { .. } | Action::PointerScroll { .. } => Capability::InputMouse,
            Action::KeyboardKey { .. }
            | Action::KeyboardText { .. }
            | Action::KeyboardShortcut { .. } => Capability::InputKeyboard,
            Action::MediaCommand { .. } => Capability::MediaControl,
            Action::PresentationCommand { .. } => Capability::PresentationControl,
            Action::AppsLaunch { .. } => Capability::AppsLaunch,
            Action::WindowsAction { .. } => Capability::WindowsControl,
            Action::ClipboardSet { .. } => Capability::ClipboardWrite,
            Action::PowerCommand { action } => match action.as_str() {
                "lock" => Capability::PowerLock,
                "sleep" => Capability::PowerSleep,
                "restart" => Capability::PowerRestart,
                "shutdown" => Capability::PowerShutdown,
                _ => Capability::PowerLock,
            },
            Action::MacroExecute { .. } => Capability::AutomationExecute,
        }
    }

    pub fn validate_bounds(&self) -> Result<(), String> {
        match self {
            Action::PointerScroll { dx, dy } => {
                if !dx.is_finite() || !dy.is_finite() {
                    return Err("Scroll deltas must be finite numbers".to_string());
                }
                if dx.abs() > 10000.0 || dy.abs() > 10000.0 {
                    return Err("Scroll deltas exceed allowable limit".to_string());
                }
            }
            Action::KeyboardText { text } => {
                if text.len() > 65536 {
                    return Err("Text payload exceeds allowable 64KB limit".to_string());
                }
            }
            Action::ClipboardSet { text } => {
                if text.len() > 1048576 {
                    return Err("Clipboard text exceeds allowable 1MB limit".to_string());
                }
            }
            Action::AppsLaunch { app_id } => {
                if app_id.trim().is_empty() || app_id.len() > 256 {
                    return Err("Invalid appId length".to_string());
                }
            }
            Action::WindowsAction {
                window_id,
                target_display,
                ..
            } => {
                if window_id.trim().is_empty() {
                    return Err("Invalid windowId".to_string());
                }
                if let Some(td) = target_display {
                    if *td > 64 {
                        return Err("Invalid target display index".to_string());
                    }
                }
            }
            _ => {}
        }
        Ok(())
    }
}
